File: 0974-subarray-sums-divisible-by-k/0974-subarray-sums-divisible-by-k.cpp
class Solution {
public:
    int subarraysDivByK(vector<int>& nums, int k) {
        int count = 0;
        int prefix_sum = 0;
        unordered_map<int, int> remainder_freq;
        remainder_freq[0] = 1;

        for (int num : nums) {
            prefix_sum += num;
            int remainder = prefix_sum % k;
            if (remainder < 0)
                remainder += k;
            count += remainder_freq[remainder];
            remainder_freq[remainder]++;
        }

        return count;
    }
};

File: 0974-subarray-sums-divisible-by-k/0974-subarray-sums-divisible-by-k.cs
public class Solution {
    public int SubarraysDivByK(int[] nums, int k) {
        int count = 0;
        int prefixSum = 0;
        Dictionary<int, int> remainderFreq = new Dictionary<int, int>();
        remainderFreq[0] = 1; // Initialize with one occurrence of 0 remainder

        foreach (int num in nums)
        {
            prefixSum += num;
            int remainder = prefixSum % k;
            if (remainder < 0) // Ensure remainder is positive
                remainder += k;
            count += remainderFreq.GetValueOrDefault(remainder);
            if (remainderFreq.ContainsKey(remainder))
                remainderFreq[remainder]++;
            else
                remainderFreq[remainder] = 1;
        }

        return count;
    }
}